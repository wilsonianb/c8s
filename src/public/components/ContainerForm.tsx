import * as React from 'react'
import { withFormik, FormikBag, FormikProps, FormikErrors, Form, Field, FieldArray, ArrayHelpers } from 'formik'
import { Manifest, ManifestOpts, EnvVar } from '../Manifest'

interface OtherProps {
  codiusVersion: string
}

const InnerForm = (props: OtherProps & FormikProps<ManifestOpts>) => {
  const { touched, errors, isSubmitting, status } = props
  return (
    <Form>
      <label>
        Docker container image URL:
        <Field type='text' name='image' placeholder='docker.io/user/image:tag' autoFocus />
        {touched.image && errors.image && <div>{errors.image}</div>}
      </label>
      <br />
      <label>
        Port:
        <Field type='number' name='port' min='1' max='65535' />
        {touched.port && errors.port && <div>{errors.port}</div>}
      </label>
      <br />
      <br />
      <FieldArray
        name='envVars'
        render={(arrayHelpers: ArrayHelpers) => (
          <div>
            {props.values.envVars.map((envVar: EnvVar, index: number) => (
              <div key={index}>
                <Field type='text' name={`envVars[${index}].name`} />
                =
                <Field type='text' name={`envVars[${index}].value`} />
                <button
                  type='button'
                  onClick={() => arrayHelpers.remove(index)}
                  title='Remove'
                >
                  -
                </button>
                <Field type='checkbox' name={`envVars[${index}].private`} checked={props.values.envVars[index].private} />
                Private
              </div>
            ))}
            <button type='button' onClick={() => arrayHelpers.push({
              name: '',
              value: '',
              private: true
            })}>
              Add environment variable
            </button>
          </div>
        )}
      />
      <br />
      <button type='submit' disabled={isSubmitting}>
        Run Container
      </button>
      <br />
      <br />

      {isSubmitting &&
        <div>
          Deploying...
        </div>
      }
      {status && status.error &&
        <div>
          {status.error}
        </div>
      }
      {status && status.containerUrl &&
        <div>
          Container is now running at <a href={status.containerUrl}>{status.containerUrl}</a>
        </div>
      }
      <br />
      {status && status.containerUrl && status.manifest &&
        <div>
          Manifest:
          <br />
          <pre>{JSON.stringify(status.manifest, null, 2)}</pre>
        </div>
      }
    </Form>
  )
}

interface MyFormProps {
  initialImage?: string
  codiusVersion: string
}

const ContainerForm = withFormik<MyFormProps, ManifestOpts>({
  displayName: 'ContainerForm',

  mapPropsToValues: (props: MyFormProps) => {
    return {
      image: props.initialImage || '',
      port: 8080,
      envVars: []
    }
  },

  // Add a custom validation function (this can be async too!)
  validate: (values: ManifestOpts) => {
    let errors: FormikErrors<ManifestOpts> = {}
    if (!values.image) {
      errors.image = 'Required'
    // } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.image)) {
    //   errors.image = 'Invalid container image'
    }
    // TODO: don't allow PORT env var
    return errors
  },

  handleSubmit: (values: ManifestOpts, { props, setStatus, setSubmitting }: FormikBag<MyFormProps, ManifestOpts>) => {
    const manifest = new Manifest(values)
    setStatus({
      error: undefined,
      containerUrl: undefined,
      manifest: undefined
    })
    return fetch('/containers', {
      headers: {
        Accept: `application/codius-v${props.codiusVersion}+json`,
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body: JSON.stringify({
        manifest: manifest.manifest,
        private: manifest.private
      })
    })
    .then(response => {
      if (response.ok) {
        return response.json()
      } else {
        throw new Error()
      }
    })
    .then(result => {
      setStatus({
        containerUrl: result.url,
        manifest: manifest.manifest
      })
    })
    .catch(error => {
      setStatus({
        error: 'Unable to deploy container'
      })
    })
    .finally(() => setSubmitting(false))
  }
})(InnerForm)

export default ContainerForm
