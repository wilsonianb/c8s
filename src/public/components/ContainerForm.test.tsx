/*!
 * @jest-environment jsdom
 */

import * as React from 'react'
import { mount, render } from 'enzyme'

import ContainerForm from './ContainerForm'

describe('<ContainerForm />', () => {
  const codiusVersion: string = '0.0.1'

  it('matches snapshot', () => {
    const wrapper = mount(<ContainerForm codiusVersion={codiusVersion} />)
    expect(wrapper).toMatchSnapshot()
  })

  // it('renders three `.foo-bar`s', () => {
  //   const wrapper = render(<ContainerForm codiusVersion={codiusVersion} />)
  //   // expect(wrapper.find('.foo-bar')).to.have.lengthOf(3)
  //   expect(wrapper.find('#child')).toHaveHTML(
  //     '<span id="child">Test</span>'
  //   )
  // })

  // it('renders the title', () => {
  //   const wrapper = render(<ContainerForm codiusVersion={codiusVersion} />)
  //   expect(wrapper.text()).toHaveText('unique')
  // })
})
