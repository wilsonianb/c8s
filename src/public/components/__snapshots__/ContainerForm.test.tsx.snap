// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContainerForm /> matches snapshot 1`] = `
<WithFormik(InnerForm)
  codiusVersion="0.0.1"
>
  <Formik
    codiusVersion="0.0.1"
    displayName="ContainerForm"
    enableReinitialize={false}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "envVars": Array [],
        "image": "",
        "port": 8080,
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <InnerForm
      codiusVersion="0.0.1"
      dirty={false}
      errors={Object {}}
      handleBlur={[Function]}
      handleChange={[Function]}
      handleReset={[Function]}
      handleSubmit={[Function]}
      initialValues={
        Object {
          "envVars": Array [],
          "image": "",
          "port": 8080,
        }
      }
      isSubmitting={false}
      isValid={false}
      isValidating={false}
      registerField={[Function]}
      resetForm={[Function]}
      setError={[Function]}
      setErrors={[Function]}
      setFieldError={[Function]}
      setFieldTouched={[Function]}
      setFieldValue={[Function]}
      setFormikState={[Function]}
      setStatus={[Function]}
      setSubmitting={[Function]}
      setTouched={[Function]}
      setValues={[Function]}
      submitCount={0}
      submitForm={[Function]}
      touched={Object {}}
      unregisterField={[Function]}
      validateField={[Function]}
      validateForm={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
      values={
        Object {
          "envVars": Array [],
          "image": "",
          "port": 8080,
        }
      }
    >
      <Form>
        <Component
          formik={
            Object {
              "dirty": false,
              "errors": Object {},
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleReset": [Function],
              "handleSubmit": [Function],
              "initialValues": Object {
                "envVars": Array [],
                "image": "",
                "port": 8080,
              },
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "registerField": [Function],
              "resetForm": [Function],
              "setError": [Function],
              "setErrors": [Function],
              "setFieldError": [Function],
              "setFieldTouched": [Function],
              "setFieldValue": [Function],
              "setFormikState": [Function],
              "setStatus": [Function],
              "setSubmitting": [Function],
              "setTouched": [Function],
              "setValues": [Function],
              "status": undefined,
              "submitCount": 0,
              "submitForm": [Function],
              "touched": Object {},
              "unregisterField": [Function],
              "validate": [Function],
              "validateField": [Function],
              "validateForm": [Function],
              "validateOnBlur": true,
              "validateOnChange": true,
              "validationSchema": undefined,
              "values": Object {
                "envVars": Array [],
                "image": "",
                "port": 8080,
              },
            }
          }
        >
          <form
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <label>
              Docker container image URL:
              <FormikConnect(FieldInner)
                autoFocus={true}
                name="image"
                placeholder="docker.io/user/image:tag"
                type="text"
              >
                <FieldInner
                  autoFocus={true}
                  formik={
                    Object {
                      "dirty": false,
                      "errors": Object {},
                      "handleBlur": [Function],
                      "handleChange": [Function],
                      "handleReset": [Function],
                      "handleSubmit": [Function],
                      "initialValues": Object {
                        "envVars": Array [],
                        "image": "",
                        "port": 8080,
                      },
                      "isSubmitting": false,
                      "isValid": false,
                      "isValidating": false,
                      "registerField": [Function],
                      "resetForm": [Function],
                      "setError": [Function],
                      "setErrors": [Function],
                      "setFieldError": [Function],
                      "setFieldTouched": [Function],
                      "setFieldValue": [Function],
                      "setFormikState": [Function],
                      "setStatus": [Function],
                      "setSubmitting": [Function],
                      "setTouched": [Function],
                      "setValues": [Function],
                      "status": undefined,
                      "submitCount": 0,
                      "submitForm": [Function],
                      "touched": Object {},
                      "unregisterField": [Function],
                      "validate": [Function],
                      "validateField": [Function],
                      "validateForm": [Function],
                      "validateOnBlur": true,
                      "validateOnChange": true,
                      "validationSchema": undefined,
                      "values": Object {
                        "envVars": Array [],
                        "image": "",
                        "port": 8080,
                      },
                    }
                  }
                  name="image"
                  placeholder="docker.io/user/image:tag"
                  type="text"
                >
                  <input
                    autoFocus={true}
                    name="image"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="docker.io/user/image:tag"
                    type="text"
                    value=""
                  />
                </FieldInner>
              </FormikConnect(FieldInner)>
            </label>
            <br />
            <label>
              Port:
              <FormikConnect(FieldInner)
                max="65535"
                min="1"
                name="port"
                type="number"
              >
                <FieldInner
                  formik={
                    Object {
                      "dirty": false,
                      "errors": Object {},
                      "handleBlur": [Function],
                      "handleChange": [Function],
                      "handleReset": [Function],
                      "handleSubmit": [Function],
                      "initialValues": Object {
                        "envVars": Array [],
                        "image": "",
                        "port": 8080,
                      },
                      "isSubmitting": false,
                      "isValid": false,
                      "isValidating": false,
                      "registerField": [Function],
                      "resetForm": [Function],
                      "setError": [Function],
                      "setErrors": [Function],
                      "setFieldError": [Function],
                      "setFieldTouched": [Function],
                      "setFieldValue": [Function],
                      "setFormikState": [Function],
                      "setStatus": [Function],
                      "setSubmitting": [Function],
                      "setTouched": [Function],
                      "setValues": [Function],
                      "status": undefined,
                      "submitCount": 0,
                      "submitForm": [Function],
                      "touched": Object {},
                      "unregisterField": [Function],
                      "validate": [Function],
                      "validateField": [Function],
                      "validateForm": [Function],
                      "validateOnBlur": true,
                      "validateOnChange": true,
                      "validationSchema": undefined,
                      "values": Object {
                        "envVars": Array [],
                        "image": "",
                        "port": 8080,
                      },
                    }
                  }
                  max="65535"
                  min="1"
                  name="port"
                  type="number"
                >
                  <input
                    max="65535"
                    min="1"
                    name="port"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="number"
                    value={8080}
                  />
                </FieldInner>
              </FormikConnect(FieldInner)>
            </label>
            <br />
            <br />
            <FormikConnect(FieldArrayInner)
              name="envVars"
              render={[Function]}
            >
              <FieldArrayInner
                formik={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialValues": Object {
                      "envVars": Array [],
                      "image": "",
                      "port": 8080,
                    },
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setError": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "status": undefined,
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validate": [Function],
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "validationSchema": undefined,
                    "values": Object {
                      "envVars": Array [],
                      "image": "",
                      "port": 8080,
                    },
                  }
                }
                name="envVars"
                render={[Function]}
                validateOnChange={true}
              >
                <div>
                  <button
                    onClick={[Function]}
                    type="button"
                  >
                    Add environment variable
                  </button>
                </div>
              </FieldArrayInner>
            </FormikConnect(FieldArrayInner)>
            <br />
            <button
              disabled={false}
              type="submit"
            >
              Run Container
            </button>
            <br />
            <br />
            <br />
          </form>
        </Component>
      </Form>
    </InnerForm>
  </Formik>
</WithFormik(InnerForm)>
`;
